layers/dhcpv4.go

```diff
@@ -0,0 +1,426 @@
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/google/gopacket"
+)
+
+var LayerTypeDHCPv4 = gopacket.RegisterLayerType(6768, gopacket.LayerTypeMetadata{"DHCPv4", gopacket.DecodeFunc(decodeDHCPv4)})
+
+var opTypes = [3]string{
+	0: "UNSPEC",
+	1: "REQUEST",
+	2: "RESPONSE",
+}
+
+type DHCPOperation uint8
+
+const (
+	DHCP_MSG_UNSPEC DHCPOperation = iota
+	DHCP_MSG_DISCOVER
+	DHCP_MSG_OFFER
+	DHCP_MSG_REQUEST
+	DHCP_MSG_DECLINE
+	DHCP_MSG_ACK
+	DHCP_MSG_NAK
+	DHCP_MSG_RELEASE
+	DHCP_MSG_INFORM
+)
+
+var messageTypes = map[DHCPOperation]string{
+	0: "UNSPEC",
+	1: "DISCOVER",
+	2: "OFFER",
+	3: "REQUEST",
+	4: "DECLINE",
+	5: "ACK",
+	6: "NAK",
+	7: "RELEASE",
+	8: "INFORM",
+}
+
+func (o DHCPOperation) String() string {
+	return opTypes[o]
+}
+
+const (
+	_ = iota
+	DHCP_MSG_REQ
+	DHCP_MSG_RES
+)
+
+var dhcpMagic uint32 = 0x63825363
+
+type DHCPv4 struct {
+	BaseLayer
+	Operation    DHCPOperation
+	HardwareType byte
+	HardwareLen  uint8
+	HardwareOpts uint8
+	Xid          uint32
+	Secs         uint16
+	Flags        uint16
+	ClientIP     net.IP
+	YourIP       net.IP
+	ServerIP     net.IP
+	GatewayIP    net.IP
+	ClientHWAddr net.HardwareAddr
+	ServerName   []byte
+	File         []byte
+	Options      []DHCPOption
+}
+
+const (
+	DHCP_OPT_REQUEST_IP     byte = iota + 50 // 0x32, 4, net.IP
+	DHCP_OPT_LEASE_TIME                      // 0x33, 4, uint32
+	DHCP_OPT_EXT_OPTS                        // 0x34, 1, 1/2/3
+	DHCP_OPT_MESSAGE_TYPE                    // 0x35, 1, 1-7
+	DHCP_OPT_SERVER_ID                       // 0x36, 4, net.IP
+	DHCP_OPT_PARAMS_REQUEST                  // 0x37, n, []byte
+	DHCP_OPT_MESSAGE                         // 0x38, n, string
+	DHCP_OPT_MAX_DHCP_SIZE                   // 0x39, 2, uint16
+	DHCP_OPT_T1                              // 0x3a, 4, uint32
+	DHCP_OPT_T2                              // 0x3b, 4, uint32
+	DHCP_OPT_CLASS_ID                        // 0x3c, n, []byte
+	DHCP_OPT_CLIENT_ID                       // 0x3d, n >=  2, []byte
+
+)
+
+const (
+	DHCP_HW_TYPE_ETHERNET byte   = 0x01
+	DHCP_FLAG_BROADCAST   uint16 = 0x80
+)
+
+// LayerType returns gopacket.LayerTypeDHCPv4
+func (d *DHCPv4) LayerType() gopacket.LayerType { return LayerTypeDHCPv4 }
+
+func (dhcp *DHCPv4) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	dhcp.Operation = DHCPOperation(data[0])
+	dhcp.HardwareType = data[1]
+	dhcp.HardwareLen = data[2]
+	dhcp.HardwareOpts = data[3]
+	dhcp.Xid = binary.BigEndian.Uint32(data[4:8])
+	dhcp.Secs = binary.BigEndian.Uint16(data[8:10])
+	dhcp.Flags = binary.BigEndian.Uint16(data[10:12])
+	dhcp.ClientIP = net.IP(data[12:16])
+	dhcp.YourIP = net.IP(data[16:20])
+	dhcp.ServerIP = net.IP(data[20:24])
+	dhcp.GatewayIP = net.IP(data[24:28])
+	dhcp.ClientHWAddr = net.HardwareAddr(data[28 : 28+dhcp.HardwareLen])
+	dhcp.ServerName = data[44:108]
+	dhcp.File = data[108:236]
+	if binary.BigEndian.Uint32(data[236:240]) != dhcpMagic {
+		return errors.New("Bad DHCP header")
+	}
+
+	if len(data) > 240 {
```

Comment:
I would prefer we keep the code structure short and do the test to return ASAP. 

Let's do here something like:

if len(data) <= 240 {
 // DHCP Packet could have no option (??)
 return nil
}